pkg/openevec/config.go

```diff
@@ -357,3 +367,93 @@ func ConfigCheck(configName string) error {
 	}
 	return nil
 }
+
+func getValStrRepr(v reflect.Value) string {
+	if v.Kind() == reflect.String {
+		return fmt.Sprintf("'%v'", v.Interface())
+	} else {
+		return fmt.Sprintf("%v", v.Interface())
+	}
+}
+
+func WriteConfig(dst reflect.Value, root string, writer io.Writer, nestLevel int) {
```

Comment:
Backward compatibility with viper :) We need yaml, plus we need to trim some paths, also for backward compatibility, in theory once we remove rest of the dependencies, we can use default functions to marshal things. And granted, we can do it on a structure level, but it'll still be through tags and reflection

Reply:
I mean, of course you can specify a specific function,

```go
func (cfg *EdenSetupArgs) WithResolvedAbsPaths(root string) *EdenSetupArgs {
  cfg.Root.Eden = utils.ResolveWithAbsPath(root, cfg.Root.Eden)
  // ...
}m
```

In case of structure tags I specify how the struct should be marshalized and presented on disk and what is the operation I need to do when I'm reading from the disk.

Reply:
I guess you won't be surprised of my specific taste in those, since I do like C++ and template methods, however if there's a readability and maintainability benefit that I don't see I'm all for changing the approach :)

Reply:
but cannot we use https://pkg.go.dev/gopkg.in/yaml.v2 to marshal yaml?

Reply:
we can, but there's still a thing with resolving paths, which can be solved either by providing function like WithResolvedAbsPaths and WithTrimmedAbsPaths where developed should specify fields, which should be resolved and then specify them to be unresolved, alternatively we can use struct tags and have generic function over fields which does that when marshalling and unmarshalling.

Reply:
I prefer struct tags, because you write them once and then you forget, whenever you need a resolved field, you just specify tag in struct, everything else is taken care of. When using functions it'll create boilerplate code with just setting specific variables and then unsetting them, granted, might be easier to read, but when you need to add a field, you need to add that field in two functions. I personally think this is not clean. I do acknowledge that this is matter of taste, so I'm fine with changing it if there are problems with code understanding and/or readability. Is there something particularly nasty about this function in your opinion? Apart from the fact that it does something very similar to standard library?

Reply:
backward compatibility - touche, but as soon as we don't have that anymore I'd prefer to resort to some standard marshalizer. the reason is simple - simplicity :smile: if we just used a standard serializer we wouldn't be having this discussion. also it makes serializing less error prone, since I don't need to have special knowledge about this function to understand how the serialization happens so there is less chance for errors.

Reply:
So you'd rather have something like this? 

```go
func ResolveAbsPath(dst reflect.Value, root string) {
	if dst.Kind() == reflect.Ptr {
		dst = dst.Elem()
	}

	for i := 0; i < dst.NumField(); i++ {
		if structTag := dst.Type().Field(i).Tag.Get("mapstructure"); structTag != "" {
			f := dst.Field(i)
			fieldType := dst.Type().Field(i)

			switch f.Kind() {
			case reflect.Struct:
				// Pass the addressable value of the struct if it can be set, else create a pointer and pass
				if f.CanAddr() {
					ResolveAbsPath(f.Addr(), root)
				} else {
					ResolveAbsPath(f, root)
				}
			case reflect.String:
				if _, ok := fieldType.Tag.Lookup("resolvepath"); ok {
					f.SetValue(strings.TrimPrefix(val, root+"/"))
				}
 			}
		}
	}
}

func (cfg* EdenSetupArgs) WriteToFile(path string) {
   cfgCopy := copy(cfg)

   ResolveAbsPath(&cfgCopy, cfg.Eden.Root)

   jsonData, err := json.Marshal(cfgCopy)
  if err != nil {
    fmt.Println("Error marshalling to JSON:", err)
    return
  }
  // .. 
}

```

Reply:
looks good - we can also put a comment that `ResolveAbsPath` is there for backward compatibility and can be removed once that's not needed anymore.

also I think if you pass param by value you can save the copy statement:
```golang
func (cfg EdenSetupArgs) WriteToFile(path string) {
```

Reply:
okay, makes sense

Reply:
Let's do this change last, because in order to marshal to yaml we need to provide structure tags to it, I'd very much like to have only one tag to marshal things. I'd like to have it in a separate PR, describing possible backward compatibility issues and all, I'll create an issue mentioning this comment rn

Reply:
sure, makes sense